### ISSUE TITLE
Possible deadlock in concurrent processing when using Process concurrency model

### ISSUE THREAD

**Issue by sorendaugaard** [jbarlow83/OCRmyPDF#758]
**Describe the bug**
Since upgrading to ocrmypdf version 11.7.3 I have seen ocrmypdf processes that seems deadlocked and hanging during processing.  

This is only reproducible when using the `ProcessPool` concurrency model. If I use the `--use-threads` argument I never see the deadlocked process.

**To Reproduce**
This is not reproducible on every run, but by running OCR multiple times I am typically able to get to lock up. 

Some files are worse than others. Unfortunately I am not able to share a file that locks up every other time, but the attached file will make it lock up if you run it multiple times in a row [multipage-jpeg.pdf](https://github.com/jbarlow83/OCRmyPDF/files/6280855/multipage-jpeg.pdf).

I use a Ruby script like this to execute the ocrmypdf multiple times and reproduce the issue:  
```
file = ARGV.first || "file-1.pdf"

COMMAND = "ocrmypdf --force-ocr --jobs 4 --language eng --optimize 3 --rotate-pages --deskew --tesseract-timeout 600 --verbose --sidecar sidecar.txt #{file} output.pdf"

puts "Running: #{COMMAND} 20 times"
20.times do |n|
        puts "ATTEMPT #{n}"
        `#{COMMAND}`
end
```

This will typically cause the process to lock up with output like this:
```
    3 input dpi = 72 x 72                                                                                                            
    3 rotation = 0°
    3 input colorspace = RGB
    3 width x height = 996px x 1284px
    3 read_images() embeds a JPEG
    3 convert done
    3 Running: ['tesseract', '-l', 'eng', '-c', 'textonly_pdf=1', PosixPath('/tmp/ocrmypdf.io.8bi6gi1x/000003_ocr.png'), '/tmp/ocrmypdf.io.8bi6gi1x/000003_ocr_tess', 'pdf', 'txt']
    2 convert
    2 imgformat = JPEG
    2 input dpi = 72 x 72
    2 rotation = 0°
    2 input colorspace = RGB
    2 width x height = 985px x 1167px
    2 read_images() embeds a JPEG
    2 convert done
    2 Running: ['tesseract', '-l', 'eng', '-c', 'textonly_pdf=1', PosixPath('/tmp/ocrmypdf.io.8bi6gi1x/000002_ocr.png'), '/tmp/ocrmypdf.io.8bi6gi1x/000002_ocr_tess', 'pdf', 'txt']
    1 convert
    1 imgformat = JPEG
    1 input dpi = 72 x 72
    1 rotation = 0°
    1 input colorspace = RGB
    1 width x height = 968px x 1285px
    1 read_images() embeds a JPEG
    1 convert done
    1 Running: ['tesseract', '-l', 'eng', '-c', 'textonly_pdf=1', PosixPath('/tmp/ocrmypdf.io.8bi6gi1x/000001_ocr.png'), '/tmp/ocrmypdf.io.8bi6gi1x/000001_ocr_tess', 'pdf', 'txt']
    2 Emplacement update
    2 Text rotation: (text, autorotate, content) -> text misalignment = (0, 0, 0) -> 0
    2 Grafting
    2 Page rotation: (content, auto) -> page = (0, 0) -> 0
    3 Emplacement update
    3 Text rotation: (text, autorotate, content) -> text misalignment = (0, 0, 0) -> 0
    3 Grafting
    3 Page rotation: (content, auto) -> page = (0, 0) -> 0
    1 Emplacement update
    1 Text rotation: (text, autorotate, content) -> text misalignment = (0, 0, 0) -> 0
    1 Grafting
    1 Page rotation: (content, auto) -> page = (0, 0) -> 0
OCR:  62%|████████████████████████████████████████████████████████████▋                                    | 2.5/4.0 [00:05<00:02,  1.75s/page
```

I will typically see 5 ocrmypdf processes hanging when it locks up:
```
logikbot  4005  3267  0 18:23 pts/1    00:00:00 /usr/bin/python3 /usr/local/bin/ocrmypdf --force-ocr --jobs 4 --language eng --optimize 3 --rot
logikbot  4024  4005  1 18:23 pts/1    00:00:01 /usr/bin/python3 /usr/local/bin/ocrmypdf --force-ocr --jobs 4 --language eng --optimize 3 --rot
logikbot  4025  4005  7 18:23 pts/1    00:00:11 /usr/bin/python3 /usr/local/bin/ocrmypdf --force-ocr --jobs 4 --language eng --optimize 3 --rot
logikbot  4027  4005  0 18:23 pts/1    00:00:01 /usr/bin/python3 /usr/local/bin/ocrmypdf --force-ocr --jobs 4 --language eng --optimize 3 --rot
logikbot  4030  4005  0 18:23 pts/1    00:00:00 /usr/bin/python3 /usr/local/bin/ocrmypdf --force-ocr --jobs 4 --language eng --optimize 3 --rot
```

When I Ctrl+C to exit the process the Traceback is:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/ocrmypdf/_sync.py", line 373, in run_pipeline
    exec_concurrent(context)
  File "/usr/local/lib/python3.6/dist-packages/ocrmypdf/_sync.py", line 285, in exec_concurrent
    task_finished=update_page,
  File "/usr/local/lib/python3.6/dist-packages/ocrmypdf/_concurrent.py", line 112, in exec_progress_pool
    for result in results:
  File "/usr/lib/python3.6/multiprocessing/pool.py", line 724, in next
    self._cond.wait(timeout)
  File "/usr/lib/python3.6/threading.py", line 295, in wait
    waiter.acquire()
```

**Expected behavior**
Ocrmypdf would proceed with processing the file. 

**Screenshots**

**System (please complete the following information):**
 - OS: Ubuntu 18 
 - Python version: Python 3.6.9
 - OCRmyPDF version: 11.7.3

**Installation**
How did you install OCRmyPDF? Did you install it from your operating system's
package manager, or using pip?

Using pip. 

**Additional context**
I have hunch this might be related to logging and the logging queues since the problem happens less often if I include the argument: `--no-progress-bar` or remove the `--verbose` argument.

**Comment by jbarlow83** [comment]
Reproduced in Docker Ubuntu 18 with the same Python, although I get it only about 1/3 of the time I run the whole ruby script.

Have you tried this with a newer Python?
So far I'm exploring the possibility of a connection to
  https://bugs.python.org/issue25942
which was resolved in Python 3.7.

Of course I can also check other versions, but I'd like to know if you've done any other tests.

**Comment by jbarlow83** [comment]
Also, if you are able to try ocrmypdf 12.0.0b1 (beta; you will have to use `pip install ocrmypdf==12.0.0b1`), there are changes to the way the executors work which might make a difference.

**Comment by jbarlow83** [comment]
I believe the issue is a long standing Python issue(s):
https://bugs.python.org/issue6721
https://bugs.python.org/issue40399

Certainly explains why verbose logging contributes. It also appears that certain improvements were made after 3.6 although the issue is still not exactly resolved.

The error occurs at leptonica.py:118, `sys.stderr.flush()`. There's an internal lock on this object that is shared across processes and that seems to be responsible for the deadlock, if the logging thread is trying to output something at the same time. An explicit `sys.stderr.close()` in `process_init()`, surprisingly, makes matters worse.

Commit 4ea9cffe may have aggravated the situation, but if it is Python issue 6721, it's still ultimately unresolved. The threading lock there is mainly defensive - in practice the lock should never be contested.

This is related to Leptonica < 1.79. On old versions we must redirect stderr to intercept the error message that it posts there (!). On >= 1.79 Leptonica added a new API at our request to allow more sane error handling, and no messing around with stderr redirection.

**Comment by sorendaugaard** [comment]
Amazingly fast turnaround time 🤩  Thank you so much @jbarlow83. 

Just so I am completely clear, the solution here is to upgrade Leptonica to a version higher than 1.79 and apply the fix in e4f69cc? or would we still occasionally see the issue with those upgrades because its a long standing Python issue?

**Comment by jbarlow83** [comment]
The recent commit, and Leptonica >=1.79, I think, should make this go away. A newer Python may also help since there are constantly little fixes to concurrency.
